File: top_hot100/0096UniqueBinarySearchTrees/num_trees.py
# coding: utf-8

class Solution:
    def numTrees(self, n):
        dp = [1, 1]
        for i in range(1, n+1):
            nums = 0
            for j in range(1, i):
                nums += dp[j-1] * dp[i-j]
            dp.append(nums)
        
        return dp[n]

File: top_hot100/0096UniqueBinarySearchTrees/num_trees.go
package numtrees

func numTrees(n int) int {
	if n <= 0 {
		return 0
	}
	dp := make([]int, n+1)
	dp[0] = 1
	dp[1] = 1
	for i := 2; i <= n; i++ {
		all := 0
		for j := 1; j <= i; j++ {
			all += dp[j-1] * dp[i-j]
		}
		dp[i] = all
	}
	return dp[n]
}
